src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs

```diff
@@ -0,0 +1,866 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Buffers;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Threading;
+
+namespace System.IO.Pipelines
+{
+    /// <summary>
+    /// Default <see cref="PipeWriter"/> and <see cref="PipeReader"/> implementation.
+    /// </summary>
+    public sealed class Pipe
+    {
+        private const int SegmentPoolSize = 16;
+
+        private static readonly Action<object> _signalReaderAwaitable = state => ((Pipe)state).ReaderCancellationRequested();
+        private static readonly Action<object> _signalWriterAwaitable = state => ((Pipe)state).WriterCancellationRequested();
+        private static readonly Action<object> _invokeCompletionCallbacks = state => ((PipeCompletionCallbacks)state).Execute();
+
+        // This sync objects protects the following state:
+        // 1. _commitHead & _commitHeadIndex
+        // 2. _length
+        // 3. _readerAwaitable & _writerAwaitable
+        private readonly object _sync = new object();
+
+        private readonly MemoryPool<byte> _pool;
+        private readonly int _minimumSegmentSize;
+        private readonly long _maximumSizeHigh;
```

Comment:
Are these the backpressure thresholds? If yes, could you rename to use the same terminology as the PipeOptions properties.